### ISSUE TITLE
[0.x] Packet loss causes freezing h264 video in firefox

### ISSUE THREAD

**Issue by groupboard** [meetecho/janus-gateway#2981]
**What version of Janus is this happening on?**
v0.12.1
**Have you tested a more recent version of Janus too?**
Yes. Latest master.

**Was this working before?**
Yes. 0.12.0

**Is there a gdb or libasan trace of the issue?**

**Additional context**
Bug was introduced by https://github.com/meetecho/janus-gateway/commit/b75fb9687f1a94ded23298d5b93fd77f5f96b589

To recreate:
- set up h264 video room
- simulate packet loss

Chrome stutters, but does not freeze. Video on firefox (received from other user) completely freezes forever after any packet loss, even after network operating normally with zero packet loss.

**Comment by groupboard** [comment]
Doing some more testing, the bug also occurs in multistream, and also with vp8 rooms. It can be recreated here in a few seconds by generating packet loss, then stopping the packet loss:

https://janus.conf.meetecho.com/videoroomtest.html

**Comment by tmatth** [comment]
> Doing some more testing, the bug also occurs in multistream, and also with vp8 rooms. It can be recreated here in a few seconds by generating packet loss, then stopping the packet loss:
> 
> https://janus.conf.meetecho.com/videoroomtest.html

@groupboard is this with `playout_ext` enabled or disabled?

**Comment by groupboard** [comment]
> playoutdelay_ext

Not sure...just the default config. Did you try the janus demos above? You'll see it immediately happening there. You can use "Network Link Conditioner" (included with XCode) on a mac to generate packet loss.

**Comment by lminiero** [comment]
We don't have a Mac. Why are you sure that commit broke it? Have you done a `git bisect`?

**Comment by groupboard** [comment]
> We don't have a Mac. 

Here are some windows tools to simulate packet loss:

https://serverfault.com/questions/302299/what-tools-are-available-on-windows-to-simulate-emulate-network-issues

>Why are you sure that commit broke it?

Because I went through each commit, and as soon as it got to that one, the bug started. Reverting that change fixes it.

Prior to this commit, the default value of 0 for the videoroom min/max delay would be inserted into the outgoing RTP packets for min/max delay, i.e. no min/max delay. However after this change, if min/max delay aren't set for the room then the values for min/max are taken from the incoming rtp packet and added to the outgoing rtp. Is that the intended behaviour?

**Comment by lminiero** [comment]
> Here are some windows tools to simulate packet loss:

I don't use windows either :stuck_out_tongue: 

> if min/max delay aren't set for the room then the values for min/max are taken from the incoming rtp packet and added to the outgoing rtp. Is that the intended behaviour?

That's always been the intended behaviour, even before @tmatth's tweak, assuming the extension was negotiated. If you don't override it, what's set in the incoming stream will be copied to the outgoing streams as well. That's why I'm not convinced that change plays any role: the commit you're excluding simply ensures that the default behaviour is "no override set".

**Comment by groupboard** [comment]
> I don't use windows either 😛

Try pulling the ethernet cable out for a couple of seconds. Or turning wifi off and back on again. You just need to make sure you don't kill the websocket connection.

> That's always been the intended behaviour, even before @tmatth's tweak, assuming the extension was negotiated. If you don't override it, what's set in the incoming stream will be copied to the outgoing streams as well. That's why I'm not convinced that change plays any role: the commit you're excluding simply ensures that the default behaviour is "no override set".

The other curious thing is that playout delay claims to be not even implemented in firefox yet:

https://bugzilla.mozilla.org/show_bug.cgi?id=1592988

Anyway, try out the test: use chrome on one browser, and firefox on another, and you'll see the firefox receiving video stream freezes immediately and permanently as soon as there is any packet loss at all, and it's fixed by removing this commit.

**Comment by lminiero** [comment]
> The other curious thing is that playout delay claims to be not even implemented in firefox yet:

If endpoints don't support it, they'll reject the extension, and it won't be used, meaning Janus will not relay the value to them. That said, I do see Firefox negotiating the feature: if they don't support it, it means they ignore the value.

Anyway, since I don't have two laptops to replicate any total loss, I chose a simpler way to replicate it, that is by telling the VideoRoom plugin not to relay media from a publisher for a few seconds, by typing this in the JavaScript console of Chrome user:

    sfutest.send({message: {request: "configure", audio: false, video: false}});

I waited a few seconds and then restored it:

    sfutest.send({message: {request: "configure", audio: true, video: true}});

and both audio and video resumed just fine on Firefox. This happened both on my local setup, and on the official demos.

**Comment by groupboard** [comment]
> Anyway, since I don't have two laptops to replicate any total loss, I

I can replicate it on one laptop, with two browser windows open (chrome and ff).

> 
> ```
> sfutest.send({message: {request: "configure", audio: false, video: false}});
> ```
> 
> I waited a few seconds and then restored it:
> 
> ```
> sfutest.send({message: {request: "configure", audio: true, video: true}});
> ```
> 
> and both audio and video resumed just fine on Firefox. This happened both on my local setup, and on the official demos.

Yes, when I do that it works fine, and the video starts playing again. But simulating packet loss replicates the problem 100% of the time, so I assume something different is happening during packet loss.

I tried turning wifi off and immediately on again, but that kills the websocket connection because it takes a ridiculous amount of time for wifi to actually reconnect again (at least on a mac). 

There seems to be a tool called tc that will simulate packet loss on linux:

https://stackoverflow.com/questions/614795/simulate-delayed-and-dropped-packets-on-linux

**Comment by lminiero** [comment]
Looks like @atoppi managed to replicate the issue, so we'll check what's wrong.

**Comment by lminiero** [comment]
I haven't been able to replicate the issue via packet loss myself, but I noticed that if you send a "configure" to the subscriber with a `min_delay` and `max_delay` of `0`, video does indeed freeze in Firefox: sending `1` gets video to flow again. Not sure if it's a bug in Firefox, since `0` is a valid value (basically meaning "render as soon as you can") and works in Chrome. I suspect something is causing `0` to be sent to the subscriber after your loss simulation, rather than not setting the extension at all (which is what `-1` means).

Can you check if doing that fixes the frozen video in your case as well? It should be enough to use a command like this in the JavaScript console:

    feeds[1].send({message: {request: "configure", mid: "1", min_delay: 1, max_delay: 1}});

In my case I only had one subscription up in the VideoRoom demo, so `feeds[1]` is the handle of that subscription; I'm also publishing audio and video, where video has mid `1`. In case either is different, just tweak the example above.

If you can confirm me that works, we'll check if there are cases when we're sending `0` when instead we shouldn't be sending anything at all (unless `0` is what the publisher themselves are setting).

**Comment by groupboard** [comment]
>Can you check if doing that fixes the frozen video in your case as well? It should be enough to use a command like this in the JavaScript console:

Yes, typing that into the firefox console immediately unfreezes the remote video from the chrome user after the packet loss, and I can also recreate the freezing by setting the min_delay and max_delay both to 0 using that same command.

**Comment by lminiero** [comment]
Just did a check, and if Firefox received `0` in the playout-delay extension from the start it's fine, even though it still freezes when tinkering a bit via API to change that value and then go back to `0`. @groupboard can you do a wireshark capture of the subscriber stream before and after the packet loss (so before and after it freezes for you), so that I can check what extensions are sent?

**Comment by groupboard** [comment]
@lminiero I just tested the online demos, and they work perfectly now. The video starts streaming normally again after the packet loss resolves.
